domain/domain.go

```diff
@@ -1413,23 +1369,33 @@ func (do *Domain) syncIndexUsageWorker(owner owner.Manager) {
 	}
 }
 
-func (do *Domain) updateStatsWorker(ctx sessionctx.Context, owner owner.Manager) {
-	defer util.Recover(metrics.LabelDomain, "updateStatsWorker", nil, false)
+func (do *Domain) statsWorker(owner owner.Manager) {
```

Comment:
So you think `loadStatsWorker` and `updateStatsWorker` data race with each other and just merge them into one?

Reply:
We call `(*Handle).execRestrictedSQL` concurrently, which uses `Handle.mu.ctx` to fetch another session from session pool to execute internal SQL.
https://github.com/pingcap/tidb/blob/bd8c71018527c720afdc4acb480909639284323f/statistics/handle/handle.go#L132-L134
However, the method may cause data race in the following code.
https://github.com/pingcap/tidb/blob/bd8c71018527c720afdc4acb480909639284323f/session/session.go#L1748-L1764
At Line 1757, we append warnings for `Handle.mu.ctx`, which causes data race.

Reply:
Yes, I think it can be fixed.

```
WARNING: DATA RACE
Write at 0x00c001ddb8f8 by goroutine 204:
  sync/atomic.CompareAndSwapInt32()
      /usr/local/go/src/runtime/race_amd64.s:334 +0xb
  sync/atomic.CompareAndSwapInt32()
      <autogenerated>:1 +0x1e
  [github.com/pingcap/tidb/sessionctx/stmtctx.(*StatementContext).AppendWarnings()](http://github.com/pingcap/tidb/sessionctx/stmtctx.(*StatementContext).AppendWarnings())
      /go/tidb/sessionctx/stmtctx/stmtctx.go:563 +0x65
  [github.com/pingcap/tidb/session.(*session).getInternalSession.func1()](http://github.com/pingcap/tidb/session.(*session).getInternalSession.func1())
      /go/tidb/session/session.go:1733 +0x449
  [github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()](http://github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor())
      /go/tidb/session/session.go:1765 +0x399
  [github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()](http://github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL())
      /go/tidb/session/session.go:1769 +0x124
  [github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQL()](http://github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQL())
      /go/tidb/statistics/handle/handle.go:132 +0x188
  [github.com/pingcap/tidb/statistics/handle.(*Handle).HandleUpdateStats()](http://github.com/pingcap/tidb/statistics/handle.(*Handle).HandleUpdateStats())
      /go/tidb/statistics/handle/update.go:731 +0x84
  [github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()](http://github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker())
      /go/tidb/domain/domain.go:1451 +0x97a
  [github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop](http://github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop)·dwrap·19()
      /go/tidb/domain/domain.go:1304 +0x71
Previous write at 0x00c001ddb8f8 by goroutine 163:
  [github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()](http://github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext())
      /go/tidb/sessionctx/variable/session.go:1019 +0x1f5
  [github.com/pingcap/tidb/executor.ResetContextOfStmt()](http://github.com/pingcap/tidb/executor.ResetContextOfStmt())
      /go/tidb/executor/executor.go:1745 +0x13d
  [github.com/pingcap/tidb/session.(*session).ExecuteStmt()](http://github.com/pingcap/tidb/session.(*session).ExecuteStmt())
      /go/tidb/session/session.go:1821 +0x396
  [github.com/pingcap/tidb/session.(*session).ExecuteInternal()](http://github.com/pingcap/tidb/session.(*session).ExecuteInternal())
      /go/tidb/session/session.go:1482 +0x484
  [github.com/pingcap/tidb/statistics/handle.(*Handle).getStatsReader()](http://github.com/pingcap/tidb/statistics/handle.(*Handle).getStatsReader())
      /go/tidb/statistics/handle/handle.go:1449 +0x2f1
  [github.com/pingcap/tidb/statistics/handle.(*Handle).getGlobalStatsReader()](http://github.com/pingcap/tidb/statistics/handle.(*Handle).getGlobalStatsReader())
      /go/tidb/statistics/handle/handle.go:1426 +0x125
  [github.com/pingcap/tidb/statistics/handle.(*Handle).LoadNeededHistograms()](http://github.com/pingcap/tidb/statistics/handle.(*Handle).LoadNeededHistograms())
      /go/tidb/statistics/handle/handle.go:597 +0x6e
  [github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()](http://github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker())
      /go/tidb/domain/domain.go:1369 +0x9ba
  [github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop](http://github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop)·dwrap·17()
      /go/tidb/domain/domain.go:1292 +0x39
```

We find the one writing is from ```loadStatsWorker``` and the other is from ```updateStatsWorker```. and they share the same ```session```. I think we can put them into one goroutine and keep them isolation.